### ISSUE TITLE
Control the preference to always show the DevTools pages on startup from DevTools

### ISSUE THREAD

**Issue by jacob314** [flutter/devtools#3918]
VSCode has custom UX to let users opt into always running the inspector. The problem is it is surfaced before users have actually used the inspector so they aren't sure whether it is valuable. We should surface UX inside the embedded devtools to opt in and out of always showing the embedded devtools.

"You have used the inspector a lot, would you like to always launch it when you open DevTools?"

"You haven't interacted with the inspector in N days, would you like to stop seeing it when you debug a flutter app?"

DevTools server should be driving whether the inspector pages are automatically shown on startup from IDEs. That way DevTools can surface the option to opt in and out of always showing the embedded DevTools on startup.

Downside: weird for an IDE setting to be tracked in DevTools rather than the individual IDES. We could mitigate this if we wanted to get fancy by surfacing UX for this in the IntelliJ and VSCode plugin settings as well as DevTools but I'm not sure the complexity would be worth it.

Bonus points:
Also offer the option of always showing DevTools on startup for command line users. I'd personally like this option but I'm biased.

**Comment by jacob314** [comment]
Fyi @DanTup @helin24 @kenzieschmoll. It would be nice to align on this. The VSCode current functionality is probably driving more inspector usage and annoying more users.

**Comment by DanTup** [comment]
> The VSCode current functionality is probably driving more inspector usage and annoying more users.

Agreed. tbh I'm thinking maybe the prompt should be removed from VS Code (but the setting it updated kept). I don't think we had the button on the debug toolbar when it was added, so now it's a simple single click to launch.

The setting would still exist if people want to opt-in to it, it's just not prompted on first run.

<img width="1197" alt="Screenshot 2022-03-25 at 07 51 26" src="https://user-images.githubusercontent.com/1078012/160077777-91518081-57ce-4c83-8cc3-13b3bb6c7edc.png">

If we wanted DevTools to drive the prompt, perhaps a fairly generic way for DevTools to send events (via the server?) to the IDE would help? This might be useful for showing release notes as well as telling the IDE to offer to run it at startup. It would keep the setting in the IDE where it's more discoverable/expected, but the logic for it in DevTools?

```
{ "event": "prompt", "type": "RELEASE_NOTES", "url": "https://..." }
{ "event": "prompt", "type": "AUTO_OPEN_SETTING" }
```

**Comment by helin24** [comment]
For IntelliJ release notes are shown pretty similarly to how embedded DevTools are shown, in a side panel. Would it not make sense in that case for the IDE to track when DevTools release notes were last shown (or some other info) and then pass this along for DevTools to decide whether or not to show release notes?

**Comment by kenzieschmoll** [comment]
> Would it not make sense in that case for the IDE to track when DevTools release notes were last shown (or some other info) and then pass this along for DevTools to decide whether or not to show release notes?

DevTools is already tracking the most recent version we have shown a user release notes for. If a newer version of release notes is available, we show them to the user and update the .devtools file. So if we want to show them in the IDE, it's probably best if DevTools sends an event to the IDEs. This way we don't have to implement this tracking logic for  3 places (IntelliJ, VS code, and standalone DevTools).

**Comment by DanTup** [comment]
> So in this scenario, DevTools would be tracking information like days since last usage of the inspector, and then deciding if we should show a prompt to the user based on this information. Then DevTools would forward the user's answer to the IDE, and the setting would be persisted by the IDE. We would also need DevTools to be aware of what the user's IDE setting for this is. We wouldn't want to prompt a user to enable the auto open setting if it is already enabled.

Since the prompt is being done by the IDE, it could just drop the notification if the setting it is already enabled. Although, if it's useful for DevTools to know, having a way for the IDE to pass some data to DevTools doesn't seem a bad idea either (we already do some of this on the querystring).

> What would the release notes display look like in the IDEs? Could we reuse how each IDE shows its own release notes or shows notes for the dart/flutter plugins?

Yeah, that's what I had in mind. For VS Code that's a native VS Code toast with a button that links out to release notes on the web.

**Comment by helin24** [comment]
I see, in that case does the IDE need to get a notification about release notes? Could DevTools instead display release notes over the inspector in the browser when it's time to show them? Then the IDE could navigate to the same URL and open the same panel

**Comment by kenzieschmoll** [comment]
DevTools currently displays them as a side panel Overlay directly in DevTools (see https://github.com/flutter/devtools/pull/3542). This could technically be enabled for the embedded version in the IDE, but I'm not sure how this would look. It would probably take up almost the entire pane but not quite, and may not feel like a native experience for a tooling window.

Here are the release notes on the Flutter website: https://docs.flutter.dev/development/tools/devtools/release-notes/release-notes-2.10.0. In DevTools we grab a minimal version of the markdown file and display that, so that we don't have the flutter website header, sidebar, etc in what we show. If we want to show release notes directly in the IDE, we would probably want to do something similar. If we want to just give users a link to view the release notes in a browser, we'd want to use a url like the example link above for 2.10.0.

**Comment by helin24** [comment]
> It may be better for the IDE to manage showing the release notes by a similar method to VS Code (showing a notification with a link to open in browser), or potentially by reusing the embedded browser in IntelliJ that we already use for the inspector (however, this may be a confusing UX for users if it is not immediately clear how to go "back" to the inspector).

Having a toast would make sense, I think. I agree having separate embedded browser tabs for DevTools inspector and DevTools release notes would be confusing. My main point of confusion was why DevTools can't display a full screen overlay of release notes (with a close button) instead of a side panel, conditioned on whether it's embedded or not, but it sounds like this would be difficult to change?

**Comment by DanTup** [comment]
For VS Code, I do think showing the same toast + launching an external browser for the release notes would probably be the best experience - it's both consistent with the extensions own release notes, and may avoid having to worry about the different possible layouts/sizes of the iframe in the embedded window.

> > Since the prompt is being done by the IDE...
> 
> I thought we wanted to have DevTools manage the prompt? Quoting the original request from Jacob:

Are we talking about managing when to show it, or actually showing it? My suggestion above was that DevTools decides _when_ to show it, but delegates the display of it to the IDE so it can look native. DevTools does have some toasts currently, and they look a little out of place in VS Code because they appear in the same location (although sometimes I've seen them chopped off) but don't look/behave the same. Whether it's worth the extra work to pass them to the IDE though, I'm not certain.

**Comment by jacob314** [comment]
> Oh I see. This would be possible, and would not be that much work. However, it sounds like this might not be the best experience for VS code? So we'd have to do this only for embedded devtools in IntelliJ, and not in VS code. Whether we want to keep the IDEs consistent with each other or use an approach more native to each specific IDE is a different question @jacob314. All options are possible.

We should strongly prefer keeping the functionality in IntelliJ and VSCode consistent in all cases where being consistent makes our work simpler or more robust and where there isn't a significant usability impact.

In this case, what matters is that users who are interested in seeing release notes are able to quickly read them and that users who do not care do not have the release notes get in the way of their job. I expect either directly embedding in DevTools or surfacing as toasts in the IDE would meet this goal.

Here is a though experiment:
Ask yourself how many user survey comments or bugs we would see filed if we chose an approach that was consistent across IDEs but not 100% aligned with a specific IDE. If the answer is 0 bugs, go with the more unified solution even if it might be a little different with the way some similar things are done in a specific IDE.

There are plenty of other items we can work on that users care deeply about that can be solved by unified solutions that users across all IDEs.

The problem is not just the upfront costs of having two implementations of the same feature but the ongoing maintenance costs from having a bigger and more fragmented API surface area. What I wouldn't want is "we just realized that we are only properly showing release notes to half our users

**Comment by kenzieschmoll** [comment]
Taking all of ^ into consideration, I'd be in favor of showing a toast in IDEs that will allow users to open release notes in the browser if they want to read them. DevTools can send a message to the IDEs indicating that this toast should be shown.

**Comment by kenzieschmoll** [comment]
Just discussed the release notes issue with Jacob briefly offline. Let's start by just having the DevTools release notes viewer enabled for the embedded app. We just need to cleanup how we handle responsive width for small screens and this should work fine. Filed https://github.com/flutter/devtools/issues/3925 to track this issue.

**Comment by helin24** [comment]
thanks for clarifying @kenzieschmoll @jacob314 !
